exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/TypeInferenceUtils.java

```diff
@@ -0,0 +1,196 @@
+import org.apache.drill.exec.planner.logical.DrillConstExecutor;
+import org.apache.drill.exec.resolver.FunctionResolver;
+import org.apache.drill.exec.resolver.FunctionResolverFactory;
+
+import java.util.List;
+import java.util.Set;
+
+public class TypeInferenceUtils {
+  public static final TypeProtos.MajorType NONE = TypeProtos.MajorType.getDefaultInstance();
+  public static final int MAX_VARCHAR_LENGTH = 65535;
+
+  private static final Set<String> setFnWithDynamicTypes = Sets.newHashSet();
+  static {
+    setFnWithDynamicTypes.add("FALTTEN");
+    setFnWithDynamicTypes.add("KVGEN");
+    setFnWithDynamicTypes.add("CONVERT_FROM");
+  }
+
+  public static RelDataType inferReturnType(final SqlOperatorBinding opBinding, final List<DrillFuncHolder> functions) {
+    if(functions == null
+        || functions.isEmpty()
+            || setFnWithDynamicTypes.contains(opBinding.getOperator().getName().toUpperCase())) {
+      return opBinding.getTypeFactory()
+          .createTypeWithNullability(opBinding.getTypeFactory().createSqlType(SqlTypeName.ANY), true);
+    }
+
+    final RelDataTypeFactory factory = opBinding.getTypeFactory();
+    final String name = opBinding.getOperator().getName().toUpperCase();
+    if(name.equals("CONCAT")) {
+    final RelDataType type = factory.createSqlType(SqlTypeName.VARCHAR, MAX_VARCHAR_LENGTH);
+      if(opBinding.getOperandType(0).isNullable()) {
+        return factory.createTypeWithNullability(type, true);
+      } else {
+        return type;
+      }
+    }
+
+    boolean allBooleanOutput = true;
+    for(DrillFuncHolder function : functions) {
+      if(function.getReturnType().getMinorType() != TypeProtos.MinorType.BIT) {
+        allBooleanOutput = false;
+        break;
+      }
+    }
+    if(allBooleanOutput) {
+      return opBinding.getTypeFactory().createSqlType(SqlTypeName.BOOLEAN);
+    }
+
+    // The following logic is just a safe play:
+    // Even if any of the input arguments has ANY type,
+    // it "might" still be possible to determine the return type based on other non-ANY types
+    for (RelDataType type : opBinding.collectOperandTypes()) {
+      if (type.getSqlTypeName() == SqlTypeName.ANY || type.getSqlTypeName() == SqlTypeName.DECIMAL) {
+        return opBinding.getTypeFactory()
+            .createTypeWithNullability(opBinding.getTypeFactory().createSqlType(SqlTypeName.ANY), true);
+      }
+    }
+
+    final List<LogicalExpression> args = Lists.newArrayList();
+    for(final RelDataType type : opBinding.collectOperandTypes()) {
+      final TypeProtos.MajorType majorType = getMajorType(type);
+      args.add(new DumbLogicalExpression(majorType));
+    }
+    final FunctionCall functionCall = new FunctionCall(opBinding.getOperator().getName(), args, ExpressionPosition.UNKNOWN);
+    final FunctionResolver functionResolver = FunctionResolverFactory.getResolver();
+    final DrillFuncHolder func = functionResolver.getBestMatch(functions, functionCall);
```

Comment:
Is it guarantee that there is a DrillFuncHolder to be found from this call? Can it possible return NULL?

Can you explain whether / where the implicit cast happen in this type infer logic?